File: desktop/src/renderer/state/marketplace-api-client.ts
// Typed fetch wrapper for the YouCoded marketplace Cloudflare Worker backend.
// Lives in renderer/ because the same React bundle runs on both desktop and Android —
// fetch calls go directly to the Worker (CORS allowlist covers both platforms).
// No IPC needed for read endpoints; write endpoints gate on token which callers supply via getToken().

export const MARKETPLACE_API_HOST = "https://wecoded-marketplace-api.destinj101.workers.dev";

export class MarketplaceApiError extends Error {
  constructor(public readonly status: number, message: string) {
    super(message);
  }
}

export interface AuthStartResponse {
  device_code: string;
  user_code: string;
  auth_url: string;
  expires_in: number;
}

// Signed-in user's profile as returned by GET /auth/me (Worker accounts Phase 1 shape).
export interface AuthMeResponse {
  id: string;
  login: string;              // GitHub login (player-tag continuity)
  display_name: string;       // account-native display name (Worker accounts Phase 1)
  avatar_url: string | null;
  handle: string | null;      // unique @handle, null until the user claims one
}

// `user` on completion: the Worker returns the profile alongside the token so
// clients can store both atomically (games player tag needs `login`).
// Optional because a Worker deployed before 2026-07 omits it.
export type AuthPollResponse =
  | { status: "pending" }
  | { status: "complete"; token: string; user?: AuthMeResponse };

export interface StatsResponse {
  generated_at: number;
  // thumbs_up / thumbs_down: marketplace overhaul (2026-08-27) — one-tap
  // feedback replaces star ratings. Optional until the Worker ships them; the
  // UI shows nothing (not "0%") when they are absent.
  plugins: Record<string, { installs: number; review_count: number; rating: number; thumbs_up?: number; thumbs_down?: number }>;
  themes: Record<string, { likes: number }>;
}

// One comment on a plugin — GET /comments/:plugin_id (marketplace overhaul,
// no Worker route yet; the workbench answers it from fixtures).
export interface CommentEntry {
  id: string;
  user_id: string;
  user_login: string;
  user_avatar_url: string;
  text: string;
  /** Unix timestamp in seconds */
  created_at: number;
}

// Shape of a single rating entry from GET /ratings/:plugin_id
export interface RatingEntry {
  /** Composite key: "<user_id>:<plugin_id>" */
  id: string;
  /** GitHub user id string, e.g. "github:123456" */
  user_id: string;
  user_login: string;
  user_avatar_url: string;
  stars: number;
  review_text: string | null;
  /** Unix timestamp in seconds */
  created_at: number;
}

export interface ListRatingsResponse {
  ratings: RatingEntry[];
}

export interface PostRatingInput {
  plugin_id: string;
  stars: 1 | 2 | 3 | 4 | 5;
  review_text?: string;
}

// ── Social graph (accounts Phase 2) — snake_case wire shapes from GET/POST /social/* ──
// A minimal public card. handle/avatar_url may be null (handle unclaimed;
// GitHub always sets an avatar, but the column is nullable so we mirror it).
export interface SocialUserCard {
  id: string;
  display_name: string;
  handle: string | null;
  avatar_url: string | null;
}
// GET /social/friends returns a bare array of these (card + presence + friends-since).
export interface FriendRow extends SocialUserCard {
  last_seen_at: number | null;
  created_at: number; // friendship row timestamp (friends-since), NOT account age
}
// GET /social/requests returns this object (both directions), NOT a bare array.
export interface RequestsPayload {
  incoming: Array<{ id: string; from: SocialUserCard; created_at: number }>;
  outgoing: Array<{ id: string; to: SocialUserCard; created_at: number }>;
}
// GET /social/blocks returns a bare array of these (owner-only view).
export interface BlockRow extends SocialUserCard {
  created_at: number; // when I blocked them
}

export interface MarketplaceApiClient {
  getStats(): Promise<StatsResponse>;
  authStart(): Promise<AuthStartResponse>;
  authPoll(deviceCode: string): Promise<AuthPollResponse>;
  /** Resolve the current session token to the signed-in user's profile. */
  authMe(): Promise<AuthMeResponse>;
  /** Update the account display name. Returns the stored value. */
  updateProfile(displayName: string): Promise<{ display_name: string }>;
  /** Claim/change the unique @handle. Returns the normalized handle. Throws MarketplaceApiError (400 invalid / 409 taken). */
  setHandle(handle: string): Promise<{ handle: string }>;
  /** Permanently delete the account (hard delete + cascade on the Worker). */
  deleteAccount(): Promise<void>;
  /** Invalidate the current session token server-side. */
  logout(): Promise<void>;
  postInstall(pluginId: string): Promise<void>;
  postRating(input: PostRatingInput): Promise<{ hidden: boolean }>;
  deleteRating(pluginId: string): Promise<void>;
  toggleThemeLike(themeId: string): Promise<{ liked: boolean }>;
  postReport(input: { rating_user_id: string; rating_plugin_id: string; reason?: string }): Promise<void>;
  /** Fetch all visible ratings for a plugin. Unauthenticated; newest-first, LIMIT 50.
   *  Pass an AbortSignal to cancel in-flight requests on unmount or refresh. */
  listRatings(pluginId: string, signal?: AbortSignal): Promise<ListRatingsResponse>;

  // ── Marketplace overhaul (2026-08-27) — thumbs + comments. NO Worker routes
  //    exist yet; see fixtures/marketplace/worker-api-mock.ts for the shapes the
  //    design assumes. Both auth'd; thumbs additionally require a prior install
  //    (same rule as ratings today, so strangers can't game the number). ──
  listComments(pluginId: string, signal?: AbortSignal): Promise<{ comments: CommentEntry[] }>;
  postComment(input: { plugin_id: string; text: string }): Promise<{ id: string }>;
  /** `null` clears the user's vote. */
  setThumb(input: { plugin_id: string; value: 'up' | 'down' | null }): Promise<void>;

  // ── Social graph (accounts Phase 2). All auth'd. The Worker 404s an unknown or
  //    blocked handle (indistinguishable — no enumeration oracle), 429s on caps,
  //    and 400s a self-request; those surface as MarketplaceApiError.status so the
  //    handler layer can forward the code to the UI. ──
  /** Exact-match handle lookup → one card. Throws 404 (unknown/blocked), 429 (cap). */
  lookupHandle(handle: string): Promise<SocialUserCard>;
  /** Send a friend request by handle. { status:"friends" } if already friends or the
   *  target had a pending request to me (mutual-accept). Throws 404/400/429. */
  sendRequest(handle: string): Promise<{ status: "pending" | "friends" }>;
  /** Both directions of my pending requests. */
  listRequests(): Promise<RequestsPayload>;
  /** Accept an incoming request (recipient only; 404 otherwise). */
  acceptRequest(id: string): Promise<void>;
  /** Decline an incoming request (silent — sender is never notified). */
  declineRequest(id: string): Promise<void>;
  /** Cancel an outgoing request I sent. */
  cancelRequest(id: string): Promise<void>;
  /** My friends (card + last_seen_at + friends-since). */
  listFriends(): Promise<FriendRow[]>;
  /** Remove a friend by their account id (silent). */
  unfriend(userId: string): Promise<void>;
  /** Block a user by account id (severs friendship + clears requests both ways). */
  block(userId: string): Promise<void>;
  /** Remove a block by account id. */
  unblock(userId: string): Promise<void>;
  /** My block list (owner-only view). */
  listBlocks(): Promise<BlockRow[]>;
  /** Full account data export (GET /auth/export) — one large JSON object. */
  exportData(): Promise<unknown>;
}

export function createMarketplaceApiClient(opts: {
  host: string;
  getToken: () => string | null;
}): MarketplaceApiClient {
  const { host, getToken } = opts;

  // signal is threaded through only for endpoints that need cancellation (e.g. listRatings).
  // Other methods can opt-in later without changing callers.
  async function request<T>(path: string, init: RequestInit & { auth?: boolean; signal?: AbortSignal } = {}): Promise<T> {
    const headers: Record<string, string> = { "Content-Type": "application/json", ...(init.headers as Record<string, string>) };
    if (init.auth) {
      const token = getToken();
      if (!token) throw new MarketplaceApiError(401, "not signed in");
      headers.Authorization = `Bearer ${token}`;
    }
    // Remove the custom 'auth' flag before passing to fetch (not a standard RequestInit field)
    const { auth: _auth, ...fetchInit } = init;
    const res = await fetch(`${host}${path}`, { ...fetchInit, headers });
    // 202 Accepted is used for poll-pending responses — treat as success with { status: "pending" }
    if (res.status === 202) return { status: "pending" } as T;
    if (!res.ok) {
      // Fix: the Worker's 400/409 errors (Hono HTTPException, e.g. "that handle is
      // taken") arrive as text/plain with the message as the raw body — NOT JSON
      // {message}. The old res.json() path threw on those and fell back to
      // res.statusText, which is EMPTY over HTTP/2 on Cloudflare, so every such
      // error reached the UI with a blank message. Read the body as text once,
      // then accept either shape: JSON {message} or the raw text itself.
      const raw = await res.text().catch(() => "");
      let message = "";
      try {
        const parsed = JSON.parse(raw) as { message?: unknown; error?: unknown };
        if (typeof parsed?.message === "string") message = parsed.message;
        // Fix: the Worker's JSON 500s use `{ ok: false, error }` (app.onError),
        // not `{ message }` — extract `error` so those don't reach the UI blank.
        else if (typeof parsed?.error === "string") message = parsed.error;
      } catch {
        // Not JSON — plain-text body; the raw text fallback below covers it.
      }
      if (!message) message = raw.trim();
      if (!message) message = `request failed (status ${res.status})`;
      throw new MarketplaceApiError(res.status, message);
    }
    // Success path: tolerate empty bodies (204 No Content endpoints) by returning {}
    return (await res.json().catch(() => ({}))) as T;
  }

  return {
    getStats: () => request<StatsResponse>("/stats", { method: "GET" }),
    authStart: () => request<AuthStartResponse>("/auth/github/start", { method: "POST" }),
    authPoll: (device_code) =>
      request<AuthPollResponse>("/auth/github/poll", {
        method: "POST",
        body: JSON.stringify({ device_code }),
      }),
    authMe: () => request<AuthMeResponse>("/auth/me", { method: "GET", auth: true }),
    updateProfile: (display_name) =>
      request<{ display_name: string }>("/auth/profile", {
        method: "PATCH",
        body: JSON.stringify({ display_name }),
        auth: true,
      }),
    setHandle: (handle) =>
      request<{ handle: string }>("/auth/handle", {
        method: "PUT",
        body: JSON.stringify({ handle }),
        auth: true,
      }),
    // Void endpoints: 204 No Content. request() tolerates empty bodies (res.json().catch → {}),
    // so the same pattern as postInstall/deleteRating works — discard the return.
    deleteAccount: async () => {
      await request("/auth/account", { method: "DELETE", auth: true });
    },
    logout: async () => {
      // WHY: best-effort revoke must not hang sign-out for minutes on a blackholed
      // connection — bound it to 5s so a dead endpoint can't freeze the UI. The
      // handler already swallows the resulting timeout/abort rejection —
      // AbortSignal.timeout() rejects with a TimeoutError DOMException, not an
      // AbortError — so offline sign-out is fine.
      await request("/auth/logout", { method: "POST", auth: true, signal: AbortSignal.timeout(5000) });
    },
    postInstall: async (plugin_id) => {
      await request("/installs", { method: "POST", body: JSON.stringify({ plugin_id }), auth: true });
    },
    postRating: (input) =>
      request<{ hidden: boolean }>("/ratings", {
        method: "POST",
        body: JSON.stringify(input),
        auth: true,
      }),
    deleteRating: async (plugin_id) => {
      await request(`/ratings/${encodeURIComponent(plugin_id)}`, { method: "DELETE", auth: true });
    },
    toggleThemeLike: (theme_id) =>
      request<{ liked: boolean }>(`/themes/${encodeURIComponent(theme_id)}/like`, {
        method: "POST",
        auth: true,
      }),
    postReport: async (input) => {
      await request("/reports", { method: "POST", body: JSON.stringify(input), auth: true });
    },
    // Unauthenticated — public read endpoint, no Authorization header needed.
    // signal allows callers to cancel mid-flight on unmount or refreshKey change.
    listRatings: (plugin_id, signal?) =>
      request<ListRatingsResponse>(`/ratings/${encodeURIComponent(plugin_id)}`, { method: "GET", signal }),
    listComments: (plugin_id, signal?) =>
      request<{ comments: CommentEntry[] }>(`/comments/${encodeURIComponent(plugin_id)}`, { method: "GET", signal }),
    postComment: (input) =>
      request<{ id: string }>("/comments", { method: "POST", body: JSON.stringify(input), auth: true }),
    setThumb: async (input) => {
      await request("/thumbs", { method: "POST", body: JSON.stringify(input), auth: true });
    },

    // ── Social graph (accounts Phase 2). All auth'd; path params URL-encoded. The
    //    action endpoints (accept/decline/cancel/unfriend/block/unblock) return
    //    { ok: true }; we discard it (async wrapper) since only success matters. ──
    lookupHandle: (handle) =>
      request<SocialUserCard>(`/social/users/${encodeURIComponent(handle)}`, { method: "GET", auth: true }),
    sendRequest: (handle) =>
      request<{ status: "pending" | "friends" }>("/social/requests", {
        method: "POST",
        body: JSON.stringify({ handle }),
        auth: true,
      }),
    listRequests: () =>
      request<RequestsPayload>("/social/requests", { method: "GET", auth: true }),
    acceptRequest: async (id) => {
      await request(`/social/requests/${encodeURIComponent(id)}/accept`, { method: "POST", auth: true });
    },
    declineRequest: async (id) => {
      await request(`/social/requests/${encodeURIComponent(id)}/decline`, { method: "POST", auth: true });
    },
    cancelRequest: async (id) => {
      await request(`/social/requests/${encodeURIComponent(id)}`, { method: "DELETE", auth: true });
    },
    listFriends: () =>
      request<FriendRow[]>("/social/friends", { method: "GET", auth: true }),
    unfriend: async (userId) => {
      await request(`/social/friends/${encodeURIComponent(userId)}`, { method: "DELETE", auth: true });
    },
    // Body is snake_case { user_id } — the Worker reads body.user_id (POST /social/blocks).
    block: async (userId) => {
      await request("/social/blocks", { method: "POST", body: JSON.stringify({ user_id: userId }), auth: true });
    },
    unblock: async (userId) => {
      await request(`/social/blocks/${encodeURIComponent(userId)}`, { method: "DELETE", auth: true });
    },
    listBlocks: () =>
      request<BlockRow[]>("/social/blocks", { method: "GET", auth: true }),
    exportData: () =>
      request<unknown>("/auth/export", { method: "GET", auth: true }),
  };
}

File: app/src/main/kotlin/com/youcoded/app/marketplace/MarketplaceApiClient.kt
package com.youcoded.app.marketplace

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Kotlin mirror of the TS MarketplaceApiClient in desktop/src/renderer/state/marketplace-api-client.ts.
 * Calls the YouCoded Cloudflare Worker backend.
 *
 * WHY OkHttp: already a project dependency (for the WebSocket bridge server), so no new deps needed.
 * WHY suspend: all callers live in SessionService's serviceScope (Dispatchers.IO), consistent with
 * how other blocking I/O in that service is dispatched.
 */

private const val TAG = "MarketplaceApiClient"
private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()

/** Mirrors the TS ApiResult<T> discriminated union — returned as JSON to the React renderer. */
sealed class ApiResult<out T> {
    data class Ok<T>(val value: T) : ApiResult<T>()
    data class Err(val status: Int, val message: String) : ApiResult<Nothing>()

    /** Serialize to a JSONObject matching the TS shape: { ok, value } or { ok, status, message } */
    fun toJson(serializeValue: (T) -> Any? = { v -> v }): JSONObject = when (this) {
        is Ok -> JSONObject().apply {
            put("ok", true)
            val v = serializeValue(value)
            when (v) {
                null               -> put("value", JSONObject.NULL)
                // Fix: JSONObject.NULL is a sentinel OBJECT, not Kotlin null — without
                // this branch the void-endpoint callers (`{ _ -> JSONObject.NULL }`)
                // fell through to else and shipped the literal string "null".
                JSONObject.NULL    -> put("value", JSONObject.NULL)
                is JSONObject      -> put("value", v)
                // Social list endpoints (friends/blocks) return a bare JSON array —
                // without this branch it would fall to toString() and ship a stringified
                // array, breaking the { ok, value } wire shape the renderer reads.
                is JSONArray       -> put("value", v)
                is Boolean         -> put("value", v)
                is Int             -> put("value", v)
                is String          -> put("value", v)
                else               -> put("value", v.toString())
            }
        }
        is Err -> JSONObject().apply {
            put("ok", false)
            put("status", status)
            put("message", message)
        }
    }
}

class MarketplaceApiClient(
    private val store: MarketplaceAuthStore,
    private val host: String = "https://wecoded-marketplace-api.destinj101.workers.dev",
) {
    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    // ── Internal helpers ─────────────────────────────────────────────────────

    private suspend fun request(
        path: String,
        method: String = "GET",
        body: JSONObject? = null,
        auth: Boolean = false,
        // WHY: lets a single call (sign-out) impose a bounded call timeout without
        // touching the shared client's defaults. Defaults to the shared `http`.
        client: OkHttpClient = http,
    ): Pair<Int, JSONObject> = withContext(Dispatchers.IO) {
        val reqBody = body?.toString()?.toRequestBody(JSON_MEDIA_TYPE)
        val builder = Request.Builder()
            .url("$host$path")
            .method(method, reqBody ?: if (method == "GET") null else "{}".toRequestBody(JSON_MEDIA_TYPE))
            .addHeader("Content-Type", "application/json")

        if (auth) {
            val token = store.getToken()
            // WHY: we don't log the token value — only whether it's present
            if (token == null) {
                return@withContext Pair(401, JSONObject().put("message", "not signed in"))
            }
            builder.addHeader("Authorization", "Bearer $token")
        }

        try {
            val resp = client.newCall(builder.build()).execute()
            val code = resp.code
            val raw = resp.body?.string() ?: "{}"
            // 202 Accepted = poll-pending — return synthetic pending body
            val json = if (code == 202) {
                JSONObject().put("status", "pending")
            } else {
                try { JSONObject(raw) } catch (_: Exception) { JSONObject() }
            }
            Pair(code, json)
        } catch (e: Exception) {
            Log.w(TAG, "HTTP $method $path failed: ${e.message}")
            Pair(0, JSONObject().put("message", e.message ?: "network error"))
        }
    }

    private fun errFromResponse(code: Int, body: JSONObject): ApiResult.Err =
        ApiResult.Err(code, body.optString("message", "HTTP $code"))

    /**
     * Raw-body variant for endpoints that return a JSON ARRAY (GET /social/friends,
     * GET /social/blocks). request() parses only a JSONObject, so a bare array would
     * be coerced to an empty object — these methods need the raw string to parse with
     * JSONArray. Mirrors request()'s auth + I/O error handling.
     */
    private suspend fun requestRaw(
        path: String,
        method: String = "GET",
        auth: Boolean = false,
    ): Pair<Int, String> = withContext(Dispatchers.IO) {
        val builder = Request.Builder()
            .url("$host$path")
            .method(method, if (method == "GET") null else "{}".toRequestBody(JSON_MEDIA_TYPE))
            .addHeader("Content-Type", "application/json")
        if (auth) {
            val token = store.getToken()
                ?: return@withContext Pair(401, "{\"message\":\"not signed in\"}")
            builder.addHeader("Authorization", "Bearer $token")
        }
        try {
            val resp = http.newCall(builder.build()).execute()
            Pair(resp.code, resp.body?.string() ?: "")
        } catch (e: Exception) {
            Log.w(TAG, "HTTP $method $path failed: ${e.message}")
            Pair(0, e.message ?: "network error")
        }
    }

    /** Extract an error message from a raw body: JSON {message} / {error}, else the raw text. */
    private fun extractMessage(raw: String, code: Int): String = try {
        val obj = JSONObject(raw)
        obj.optString("message").ifEmpty { obj.optString("error").ifEmpty { raw.trim().ifEmpty { "HTTP $code" } } }
    } catch (_: Exception) {
        raw.trim().ifEmpty { "HTTP $code" }
    }

    // ── Public API (mirrors TS client method by method) ──────────────────────

    /** POST /auth/github/start — initiates device-code OAuth flow */
    suspend fun authStart(): ApiResult<JSONObject> {
        val (code, body) = request("/auth/github/start", method = "POST")
        return if (code == 200) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /**
     * POST /auth/github/poll — polls for token.
     * Returns either { status: "pending" } (202) or { status: "complete", token: "..." } (200).
     * WHY: on complete, the caller in SessionService saves the token to the store.
     */
    suspend fun authPoll(deviceCode: String): ApiResult<JSONObject> {
        val (code, body) = request(
            "/auth/github/poll",
            method = "POST",
            body = JSONObject().put("device_code", deviceCode),
        )
        return if (code == 200 || code == 202) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /installs — records an install. Requires token. */
    suspend fun postInstall(pluginId: String): ApiResult<Unit> {
        val (code, body) = request(
            "/installs",
            method = "POST",
            body = JSONObject().put("plugin_id", pluginId),
            auth = true,
        )
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** POST /ratings — submit or update a rating. Requires token. */
    suspend fun postRating(
        pluginId: String,
        stars: Int,
        reviewText: String?,
    ): ApiResult<JSONObject> {
        val payload = JSONObject().apply {
            put("plugin_id", pluginId)
            put("stars", stars)
            if (!reviewText.isNullOrEmpty()) put("review_text", reviewText)
        }
        val (code, body) = request("/ratings", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** DELETE /ratings/:pluginId — remove the caller's rating. Requires token. */
    suspend fun deleteRating(pluginId: String): ApiResult<Unit> {
        val encoded = URLEncoder.encode(pluginId, "UTF-8")
        val (code, body) = request("/ratings/$encoded", method = "DELETE", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** POST /themes/:themeId/like — toggle like. Requires token. */
    suspend fun toggleThemeLike(themeId: String): ApiResult<JSONObject> {
        val encoded = URLEncoder.encode(themeId, "UTF-8")
        val (code, body) = request("/themes/$encoded/like", method = "POST", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    // ── Marketplace feedback (overhaul §1.7) — thumbs + comments ──

    /** POST /thumbs — "up" / "down" / null (clear). Requires token; 403 without a
     *  prior install. The response carries the plugin's NEW totals so the React
     *  UI can move the number without re-fetching /stats. */
    suspend fun setThumb(pluginId: String, value: String?): ApiResult<JSONObject> {
        val payload = JSONObject().apply {
            put("plugin_id", pluginId)
            put("value", value ?: JSONObject.NULL)
        }
        val (code, body) = request("/thumbs", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** GET /thumbs/{id} — the caller's own vote, or null. URL-encoded like
     *  toggleThemeLike: a bundle member's id is `<bundle>/<name>`, and an
     *  unencoded slash would address a different path. */
    suspend fun getThumb(pluginId: String): ApiResult<JSONObject> {
        val encoded = URLEncoder.encode(pluginId, "UTF-8")
        val (code, body) = request("/thumbs/$encoded", method = "GET", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /comments — requires token, no install needed (asking a question
     *  BEFORE installing is the point). */
    suspend fun postComment(pluginId: String, text: String): ApiResult<JSONObject> {
        val payload = JSONObject().apply {
            put("plugin_id", pluginId)
            put("text", text)
        }
        val (code, body) = request("/comments", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    // ── Account endpoints (Worker accounts Phase 1) — all auth'd, small JSON bodies ──
    // These mirror the desktop MarketplaceApiClient account methods. The wire shape
    // returned to React (via ApiResult.toJson) is identical to marketplace:rate etc.

    /** GET /auth/me — full account profile. Used by the account:user lazy heal. */
    suspend fun authMe(): ApiResult<JSONObject> {
        val (code, body) = request("/auth/me", method = "GET", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** PATCH /auth/profile — update display name. Returns { display_name }. */
    suspend fun updateProfile(displayName: String): ApiResult<JSONObject> {
        val (code, body) = request(
            "/auth/profile",
            method = "PATCH",
            body = JSONObject().put("display_name", displayName),
            auth = true,
        )
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** PUT /auth/handle — claim/change unique @handle. Returns { handle }. */
    suspend fun setHandle(handle: String): ApiResult<JSONObject> {
        val (code, body) = request(
            "/auth/handle",
            method = "PUT",
            body = JSONObject().put("handle", handle),
            auth = true,
        )
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** DELETE /auth/account — permanent hard-delete (Worker cascades all rows). */
    suspend fun deleteAccount(): ApiResult<JSONObject> {
        // 204/empty body comes back as JSONObject() from request() → Ok (2xx). Caller clears local session.
        val (code, body) = request("/auth/account", method = "DELETE", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /auth/logout — server-side session revocation (best-effort on sign-out). */
    suspend fun logout(): ApiResult<JSONObject> {
        // Sign-out is best-effort server-side revocation: a hung network must not
        // hold the UI. Desktop uses a 5s AbortSignal; this is the Kotlin mirror.
        val logoutClient = http.newBuilder().callTimeout(5, TimeUnit.SECONDS).build()
        val (code, body) = request("/auth/logout", method = "POST", auth = true, client = logoutClient)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /reports — report a rating. Requires token. */
    suspend fun postReport(
        ratingUserId: String,
        ratingPluginId: String,
        reason: String?,
    ): ApiResult<Unit> {
        val payload = JSONObject().apply {
            put("rating_user_id", ratingUserId)
            put("rating_plugin_id", ratingPluginId)
            if (!reason.isNullOrEmpty()) put("reason", reason)
        }
        val (code, body) = request("/reports", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    // ── Account data export (accounts Phase 2) ───────────────────────────────
    /** GET /auth/export — the full account data dump (one big JSON object). */
    suspend fun exportData(): ApiResult<JSONObject> {
        val (code, body) = request("/auth/export", method = "GET", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    // ── Social graph (accounts Phase 2) — all auth'd, mirror the desktop client.
    //    Object-returning endpoints reuse request(); the two LIST endpoints return
    //    a bare JSON array, so they use requestRaw() + JSONArray. Path params are
    //    URL-encoded (same convention as deleteRating above). ──

    /** GET /social/users/:handle — one card. 404 = unknown OR blocked (no oracle); 429 = cap. */
    suspend fun lookupHandle(handle: String): ApiResult<JSONObject> {
        val (code, body) = request("/social/users/${URLEncoder.encode(handle, "UTF-8")}", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /social/requests — body { handle }. Returns { status: "pending"|"friends" }. 404/400/429. */
    suspend fun sendRequest(handle: String): ApiResult<JSONObject> {
        val (code, body) = request(
            "/social/requests",
            method = "POST",
            body = JSONObject().put("handle", handle),
            auth = true,
        )
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** GET /social/requests — { incoming: [...], outgoing: [...] }. */
    suspend fun listRequests(): ApiResult<JSONObject> {
        val (code, body) = request("/social/requests", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /social/requests/:id/accept. */
    suspend fun acceptRequest(id: String): ApiResult<Unit> {
        val (code, body) = request("/social/requests/${URLEncoder.encode(id, "UTF-8")}/accept", method = "POST", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** POST /social/requests/:id/decline. */
    suspend fun declineRequest(id: String): ApiResult<Unit> {
        val (code, body) = request("/social/requests/${URLEncoder.encode(id, "UTF-8")}/decline", method = "POST", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** DELETE /social/requests/:id — cancel an outgoing request. */
    suspend fun cancelRequest(id: String): ApiResult<Unit> {
        val (code, body) = request("/social/requests/${URLEncoder.encode(id, "UTF-8")}", method = "DELETE", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** GET /social/friends — bare array of FriendRow. */
    suspend fun listFriends(): ApiResult<JSONArray> {
        val (code, raw) = requestRaw("/social/friends", auth = true)
        return if (code in 200..299) {
            // Fix: a 2xx with an unparseable body is a FAULT, not an empty list —
            // an empty-array fallback would render "no friends" and mask the bug.
            try { ApiResult.Ok(JSONArray(raw)) } catch (_: Exception) { ApiResult.Err(code, "malformed response") }
        } else ApiResult.Err(code, extractMessage(raw, code))
    }

    /** DELETE /social/friends/:userId. */
    suspend fun unfriend(userId: String): ApiResult<Unit> {
        val (code, body) = request("/social/friends/${URLEncoder.encode(userId, "UTF-8")}", method = "DELETE", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** POST /social/blocks — body { user_id }. Severs friendship + clears requests both ways. */
    suspend fun block(userId: String): ApiResult<Unit> {
        val (code, body) = request(
            "/social/blocks",
            method = "POST",
            body = JSONObject().put("user_id", userId),
            auth = true,
        )
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** DELETE /social/blocks/:userId. */
    suspend fun unblock(userId: String): ApiResult<Unit> {
        val (code, body) = request("/social/blocks/${URLEncoder.encode(userId, "UTF-8")}", method = "DELETE", auth = true)
        return if (code in 200..299) ApiResult.Ok(Unit) else errFromResponse(code, body)
    }

    /** GET /social/blocks — bare array of BlockRow (owner-only view). */
    suspend fun listBlocks(): ApiResult<JSONArray> {
        val (code, raw) = requestRaw("/social/blocks", auth = true)
        return if (code in 200..299) {
            // Fix: 2xx + unparseable body = fault, not an empty block list (see listFriends).
            try { ApiResult.Ok(JSONArray(raw)) } catch (_: Exception) { ApiResult.Err(code, "malformed response") }
        } else ApiResult.Err(code, extractMessage(raw, code))
    }

    // ── Games arcade (spec §6.1) ─ mirrors the three desktop client methods in
    //    marketplace-api-client.ts (gameScores / gameBoard / submitGameScore).
    //    All three are auth'd and return a JSON OBJECT, so plain request() serves
    //    them ─ no requestRaw()/JSONArray branch is needed here.
    //
    //    Scores cross this boundary as RAW NUMBERS. How a game WORDS a score
    //    ("31 pipes", "12,480") lives in that game's entry in the renderer's
    //    game-registry.ts, so adding a game never touches this file. ──

    /** GET /games/scores ─ my own best in every solo game I have played, keyed by
     *  game id: { "flappy": { best, best_at, runs } }. `{}` when nothing has been
     *  played, which is a legitimate empty result and not an error. */
    suspend fun gameScores(): ApiResult<JSONObject> {
        val (code, body) = request("/games/scores", method = "GET", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** GET /games/scores/:game ─ one game's friends leaderboard,
     *  { game, you, entries }. URL-encoded like the other path params above. */
    suspend fun gameBoard(game: String): ApiResult<JSONObject> {
        val (code, body) = request("/games/scores/${URLEncoder.encode(game, "UTF-8")}", method = "GET", auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }

    /** POST /games/scores ─ publish a finished run. Returns
     *  { ok, best, best_at, runs, is_best }; `is_best` is the one fact the
     *  end-of-run screen cannot work out for itself. */
    suspend fun submitGameScore(game: String, score: Int): ApiResult<JSONObject> {
        val payload = JSONObject().apply {
            put("game", game)
            put("score", score)
        }
        val (code, body) = request("/games/scores", method = "POST", body = payload, auth = true)
        return if (code in 200..299) ApiResult.Ok(body) else errFromResponse(code, body)
    }
}
